### ISSUE TITLE
Hard to convert Complex TypeScript Interfaces to Zod Schema with full typing

### ISSUE THREAD

**Issue by alan890104** [colinhacks/zod#3886]
Hi everyone,

I'm working on a TypeScript project and need to convert a set of complex interfaces into a Zod schema for runtime validation. Here's a simplified example of the interfaces I'm working with:

```typescript
export type WithdrawMode = 'Single' | 'Balanced';

interface BaseWithdraw {
  pool: Address;
  removeLpAmount: bigint;
  queryId?: bigint;
  slippageTolerance?: Slippage;
  minAmountOuts?: IAllocation[] | IAllocation;
  extraPayload?: ExtraPayload;
}

interface NextWithdraw {
  pool: Address;
  mode: WithdrawMode;
  withdrawAsset?: Asset;
  removeLpAmount?: bigint;
}

interface SingleWithdrawBase extends BaseWithdraw {
  mode: 'Single';
}

export interface SingleWithdrawWithNext extends SingleWithdrawBase {
  nextWithdraw: NextWithdraw;
  withdrawAsset?: never;
}

interface SingleWithdrawNoNext extends SingleWithdrawBase {
  nextWithdraw?: NextWithdraw;
  withdrawAsset: Asset;
}

type SingleWithdrawParams = SingleWithdrawWithNext | SingleWithdrawNoNext;

interface BalancedWithdrawParams extends BaseWithdraw {
  mode: 'Balanced';
  nextWithdraw?: NextWithdraw;
}

export type WithdrawParams = SingleWithdrawParams | BalancedWithdrawParams;
```

My goal is to create Zod schemas for these types, especially for `WithdrawParams`, which involves unions and nested objects. I'm unsure how to handle the more complex parts, like:

- Union types (`SingleWithdrawParams` and `BalancedWithdrawParams`)
- Conditional fields (`withdrawAsset` and `nextWithdraw`)
- Optional fields and nested objects (`NextWithdraw`)

Additionally, I have a specific requirement for the `SingleWithdrawParams` union:

For `SingleWithdrawWithNext`, I want the user to fill in only the `nextWithdraw` field, and not the `withdrawAsset` field.
For `SingleWithdrawNoNext`, I want the user to fill in only the `withdrawAsset` field, and not the `nextWithdraw` field.
These fields are mutually exclusive.

Could anyone provide guidance or examples of how to translate these interfaces into Zod schemas? Any advice or resources would be greatly appreciated!

Thank you!

_Originally posted by @alan890104 in https://github.com/colinhacks/zod/discussions/3885_

**Comment by alan890104** [comment]
Fore sure!!

**Comment by kockar96** [comment]
Were you intentionally defining a discriminated union in this example?
With having key "mode" as discriminator?

**Comment by alan890104** [comment]
Yes that is a discriminator, but it is really hard for building `zod.lazy` type, appreciate your help!

**Comment by kockar96** [comment]
Sorry for waiting. I was really busy in the last couple of weeks!
Let me know if this works for you! If not we can reiterate over! 😄

import z from 'zod';

const SingleWithdrawModeSchema = z.literal('Single');
type SingleWithdrawMode = z.infer<typeof SingleWithdrawModeSchema>;

const BalancedWithdrawModeSchema = z.literal('Balanced');
type BalancedWithdrawMode = z.infer<typeof BalancedWithdrawModeSchema>;

const WithdrawModeSchema = z.union([
  SingleWithdrawModeSchema,
  BalancedWithdrawModeSchema,
]);
type WithdrawMode = z.infer<typeof withdrawModeSchema>;

const AddressSchema = z.object({
  flatNumber: z.number(),
  address: z.string(),
});

const IdSchema = z.string().uuid(); // As example, maybe not uuid in your case
const SlippageSchema = z.string(); // What's this?
const IAllocationSchema = z.object({
  asset: AddressSchema,
  amount: z.number(),
});
const ExtraPayloadSchema = z.number();

const BaseWithdrawSchema = z.object({
  pool: AddressSchema,
  removeLpAmount: z.number(),
  queryId: IdSchema.optional(),
  slippageTolerance: SlippageSchema.optional(),
  minAmountOuts: z.union([IAllocationSchema.array(), IAllocationSchema]),
  extraPayload: ExtraPayloadSchema.optional(),
});
type BaseWithdraw = z.infer<typeof BaseWithdrawSchema>;

const NextWithdrawSchema = z.object({
  pool: AddressSchema,
  mode: withdrawModeSchema,
  withdrawAsset: AddressSchema.optional(),
  removeLpAmount: z.bigint().optional(),
});

type NextWithdraw = z.infer<typeof NextWithdrawSchema>;

const SingleWithdrawBaseSchema = BaseWithdrawSchema.merge(
  z.object({
    mode: SingleWithdrawModeSchema,
  }),
);
// Or you can do
// const SingleWithdrawBaseSchema = z.object({
//   ...BaseWithdrawSchema.shape,
//   mode: SingleWithdrawModeSchema,
// });

type SingleWithdrawBase = z.infer<typeof SingleWithdrawBaseSchema>;

const NextWithdrawSchema = z.number();

const SingleWithdrawWithNextSchema = SingleWithdrawBaseSchema.merge(
  z.object({
    withdrawAsset: NextWithdrawSchema,
  }),
).omit({
  withdrawAsset: true,
}); // Not sure if you need never, but if you do, let me know so I can adjust the schema

export type SingleWithdrawWithNext = z.infer<
  typeof SingleWithdrawWithNextSchema
>;

const SingleWithdrawNoNextSchema = SingleWithdrawBaseSchema.merge(
  z.object({
    withdrawAsset: AddressSchema,
    nextWithdraw: NextWithdrawSchema.optional(),
  }),
);

type SingleWithdrawNoNext = z.infer<typeof SingleWithdrawNoNextSchema>;

const SingleWithdrawParamsSchema = z.union([
  SingleWithdrawWithNextSchema,
  SingleWithdrawNoNextSchema,
]);

type SingleWithdrawParams = z.infer<typeof SingleWithdrawParamsSchema>;


const BalancedWithdrawParamsSchema = z.object({
  ...BaseWithdrawSchema.shape,
  mode: BalancedWithdrawModeSchema,
  nextWithdraw: NextWithdrawSchema.optional(),
});

type BalancedWithdrawParams = z.infer<typeof BalancedWithdrawParamsSchema>;


const WithdrawParamsSchema = z.union([SingleWithdrawParamsSchema, BalancedWithdrawParamsSchema]);

export type WithdrawParams = z.infer<typeof WithdrawParamsSchema>;

**Comment by alan890104** [comment]
Wow! Lemme try it over!

**Comment by kockar96** [comment]
Note: I updated the code to be in code block as GitHub doesn't play well with <> 😄
